File: workers/research/src/black_book_research/relevance/decisions.py
"""Include / exclude / supporting-context decision logic (BB-040)."""

from __future__ import annotations

from black_book_constitution import evaluate_relevance, load_product_constitution

from .gates import gate_failed
from .types import RelevanceCandidateInput, RelevanceDecision, RelevanceGateResult


def _is_weak_only_signal(candidate: RelevanceCandidateInput) -> bool:
    return (
        candidate.signals.strength == "weak"
        and candidate.signals.outcome == "candidate_only"
    )


def derive_provisional_decision(
    *,
    candidate: RelevanceCandidateInput,
    composite_score: float,
    gates: tuple[RelevanceGateResult, ...],
    policy: dict | None = None,
) -> tuple[RelevanceDecision, str | None]:
    document = policy if policy is not None else load_product_constitution()
    thresholds = document["relevanceThresholds"]

    duplicate_gate = gate_failed(gates, "duplicate")
    if duplicate_gate:
        return "exclude", duplicate_gate.reason

    signal_gate = gate_failed(gates, "signal_present")
    if signal_gate:
        return "exclude", signal_gate.reason

    negative_gate = gate_failed(gates, "negative_only")
    if negative_gate:
        return "exclude", negative_gate.reason

    capped_score = composite_score
    if _is_weak_only_signal(candidate):
        capped_score = min(capped_score, thresholds["weakSignalIndependentCeiling"])

    if capped_score < thresholds["excludeBelow"]:
        failed = next((gate for gate in gates if not gate.passed), None)
        return (
            "exclude",
            failed.reason
            if failed
            else (
                f"Composite score {capped_score:.2f} is below exclusion threshold "
                f"{thresholds['excludeBelow']}."
            ),
        )

    weak_gate = gate_failed(gates, "weak_signal_independent")
    can_include = (
        capped_score >= thresholds["includeMinimum"]
        and candidate.signals.outcome == "promotable"
        and weak_gate is None
    )
    if can_include:
        return "include", None

    if capped_score >= thresholds["supportingContextMinimum"]:
        return "supporting_context", None

    return (
        "exclude",
        f"Composite score {capped_score:.2f} did not reach supporting-context minimum.",
    )


def evaluate_decision_passes(
    composite_score: float,
    decision: RelevanceDecision,
    policy: dict | None = None,
) -> bool:
    return bool(evaluate_relevance(composite_score, decision, policy)["passes"])

File: packages/domain/src/relevance/decisions.ts
/**
 * Include / exclude / supporting-context decision logic (BB-040).
 */
import type { ProductConstitution, RelevanceDecision } from '@black-book/schemas';
import { evaluateRelevance, loadProductConstitution } from '@black-book/schemas';
import type { DiscoveryCandidateRecord } from '../discovery/types.js';
import { firstFailedGate, gateFailed } from './gates.js';
import type { RelevanceGateResult } from './types.js';

export type ProvisionalDecisionInput = {
  readonly candidate: DiscoveryCandidateRecord;
  readonly compositeScore: number;
  readonly gates: readonly RelevanceGateResult[];
  readonly policy?: ProductConstitution;
};

export type ProvisionalDecisionResult = {
  readonly decision: RelevanceDecision;
  readonly exclusionReason?: string;
};

function isWeakOnlySignal(candidate: DiscoveryCandidateRecord): boolean {
  return candidate.signals.strength === 'weak' && candidate.signals.outcome === 'candidate_only';
}

/** Derive provisional decision from gates and constitution thresholds. */
export function deriveProvisionalDecision(
  input: ProvisionalDecisionInput,
): ProvisionalDecisionResult {
  const policy = input.policy ?? loadProductConstitution();
  const thresholds = policy.relevanceThresholds;
  const { candidate, compositeScore, gates } = input;

  const duplicateGate = gateFailed(gates, 'duplicate');
  if (duplicateGate) {
    return { decision: 'exclude', exclusionReason: duplicateGate.reason };
  }

  const signalGate = gateFailed(gates, 'signal_present');
  if (signalGate) {
    return { decision: 'exclude', exclusionReason: signalGate.reason };
  }

  const negativeGate = gateFailed(gates, 'negative_only');
  if (negativeGate) {
    return { decision: 'exclude', exclusionReason: negativeGate.reason };
  }

  let cappedScore = compositeScore;
  if (isWeakOnlySignal(candidate)) {
    cappedScore = Math.min(cappedScore, thresholds.weakSignalIndependentCeiling);
  }

  if (cappedScore < thresholds.excludeBelow) {
    const failed = firstFailedGate(gates);
    return {
      decision: 'exclude',
      exclusionReason:
        failed?.reason ??
        `Composite score ${cappedScore.toFixed(2)} is below exclusion threshold ${thresholds.excludeBelow}.`,
    };
  }

  const weakGate = gateFailed(gates, 'weak_signal_independent');
  const canInclude =
    cappedScore >= thresholds.includeMinimum &&
    candidate.signals.outcome === 'promotable' &&
    !weakGate;

  if (canInclude) {
    return { decision: 'include' };
  }

  if (cappedScore >= thresholds.supportingContextMinimum) {
    return { decision: 'supporting_context' };
  }

  return {
    decision: 'exclude',
    exclusionReason: `Composite score ${cappedScore.toFixed(2)} did not reach supporting-context minimum.`,
  };
}

/** Evaluate final pass/fail for the chosen decision against constitution thresholds. */
export function evaluateDecisionPasses(
  compositeScore: number,
  decision: RelevanceDecision,
  policy?: ProductConstitution,
): boolean {
  return evaluateRelevance(compositeScore, decision, policy).passes;
}
